File: kata/its_almost_compression/its_almost_compression.py
import sys
from itertools import groupby

SAMPLE_INPUT = """\
aaaaaiiiixqvsm
rrdkuuuuyyyrrrrgghc
xhzzzccccvvsssqppc
jbiiiulllllvvvvtttttxxxxxs""".splitlines()

SAMPLE_OUTPUT = """\
5a4ixqvsm
rrdk4u3y4rgghc
xh3z4cvv3sqppc
jb3iu5l4v5t5xs""".splitlines()


def handle_group(group):
    x, g = group
    size = len(list(g))
    if size == 1:
        return x
    elif size == 2:
        return x * 2
    else:
        return str(size) + x


def solve(line: str) -> str:
    return ''.join(map(handle_group, groupby(line.strip())))


def main():
    assert SAMPLE_OUTPUT == [solve(line) for line in SAMPLE_INPUT]
    for line in sys.stdin:
        print(solve(line))


if __name__ == '__main__':
    main()

File: kata/its_almost_compression/its_almost_compression.rb
# Translated from Python to Ruby by ChatGPT

SAMPLE_INPUT = %w[
  aaaaaiiiixqvsm
  rrdkuuuuyyyrrrrgghc
  xhzzzccccvvsssqppc
  jbiiiulllllvvvvtttttxxxxxs
]

SAMPLE_OUTPUT = %w[
  5a4ixqvsm
  rrdk4u3y4rgghc
  xh3z4cvv3sqppc
  jb3iu5l4v5t5xs
]

def handle_group(char, group)
  size = group.size
  if size == 1
    char
  elsif size == 2
    char * 2
  else
    "#{size}#{char}"
  end
end

def solve(line)
  grouped = line.chars.chunk_while { |i, j| i == j }.to_a
  grouped.map { |group| handle_group(group[0], group) }.join
end

def main
  raise 'Assertion failed' unless SAMPLE_OUTPUT == SAMPLE_INPUT.map { |line| solve(line) }

  ARGF.each_line do |line|
    puts solve(line.chomp)
  end
end

main
